proposals/extension-interfaces.md

```diff
@@ -0,0 +1,194 @@
+# Extension Interface Implementation
+
+This covers open semantic questions and unresolved problems in extension interface implementation.
+
+## What
+
+The basic concept for extension interfaces is to allow additional interfaces to be implemented on existing types. For instance,
+```C#
+extension EquatableUri for Uri : IEquatable<Uri> {
+    public bool Equals(Uri other) => this == other;
+}
+```
+This extension would “implement” the interface IEquatable<T> for Uri. The C# compiler would, in an appropriate context, “attach” this implementation to a given instance of System.Uri to satisfy an implementation requirement.
+This is a very high-level explanation. The exact semantics and requirements of the feature will be detailed below.
+
+## Why
+The basic idea and inspiration traces back to Wadler’s paper [1] on a feature called “type classes.” The basic goal of the feature is to allow the description of arbitrary sets of functions (interfaces) and arbitrary implementations of those sets (interface implementations). C# already allows some of those features, but the current design has severe limitations in two areas:
```

Comment:
Should there be a link to the paper `[1]`?

Reply:
Yup, endnotes got lost somehow. It's https://dl.acm.org/doi/pdf/10.1145/75277.75283